DDF-1612 Add error reporting for config editing in Admin UI

brendan-hofmann opened the pull request:
@djblue @AzGoalie @harrison-tarr @andrewkfiedler @stustison 

@clockard will hero

There are a couple of things you can do to manually try this out. In chrome, you can open a config to edit in the admin ui, then hit `/logout` in a separate  tab and try to save the configuration. You should get an error because you no longer have admin privileges to hit jolokia. You can also hook up a debugger and interfere on the backend to cause an exception to be thrown when the config is saved. This has the additional benefit of showing off the new spinner while the backend is suspended.

If any errors occur, they look something like this (`Forbidden` gets a special message):
![screen shot 2015-11-12 at 14 37 27](https://cloud.githubusercontent.com/assets/8660335/11132072/23ef9168-894b-11e5-8a7c-7a9a34e0c307.png)



<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf/314)
<!-- Reviewable:end -->

AzGoalie commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/ui/admin/api/ConfigurationAdminMBean.java, line 208 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2JqUKfgWL2z_uPTPI3)** ([raw file](https://github.com/codice/ddf/blob/1618a092deb6263b5a6306ad4e13f66be6194b57/platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/ui/admin/api/ConfigurationAdminMBean.java#L208)):</sup>
Should we add something to the @return java doc about this or is this good enough?
<!-- Reviewable comment -K2JqUKfgWL2z_uPTPI4:-2071895241 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/ui/admin/api/ConfigurationAdminMBean.java, line 208 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2JqUKfgWL2z_uPTPI3)** ([raw file](https://github.com/codice/ddf/blob/1618a092deb6263b5a6306ad4e13f66be6194b57/platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/ui/admin/api/ConfigurationAdminMBean.java#L208)):</sup>
I'm undecided. It's not a real return value in the sense that there's nothing that makes sense to return for this method. It's only there to force jolokia to treat this call as synchronous instead of returning immediately.

I'm open to suggestions for a better return value  that would make sense to javadoc, though.
<!-- Reviewable comment -K2Jr5QddXIQDSkk5e_7:-187199013 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

AzGoalie commented:
<sup>**[platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/ui/admin/api/ConfigurationAdminMBean.java, line 208 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2JqUKfgWL2z_uPTPI3)** ([raw file](https://github.com/codice/ddf/blob/1618a092deb6263b5a6306ad4e13f66be6194b57/platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/ui/admin/api/ConfigurationAdminMBean.java#L208)):</sup>
I can't think of a better way, so it should be fine.
<!-- Reviewable comment -K2h9kYtLKAoROR-wLd8:-77929659 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

AzGoalie commented:
:+1:

AzGoalie commented:
:+1:

clockard commented on platform/admin/ui/src/main/webapp/js/views/configuration/ConfigurationEdit.view.js:
```diff
@@ -255,9 +257,29 @@ define([
          */
-        submitData: function() {
+        submitData: function () {
+            var spinner = new Spinner(define({
```
Is there a reason we don't just use the spinnerConfig.js here?

brendan-hofmann replied:
It's tweaked a bit from the defaults in spinnerConfig.js, mainly changed the color to contrast better with the modal background. I could still use a modified spinnerConfig.js here, if that wouldn't be too confusing. Thoughts?

clockard replied:
Might be nice if we could reuse the spinnerConfig.js here and just change the color or whatever else you need to here.

brendan-hofmann replied:
Yeah, that sounds better, I'll do that instead.

pklinef replied:
Are you still going to use the default values and then extend them with your changes?

stustison commented on platform/admin/ui/src/main/webapp/js/views/configuration/ConfigurationEdit.view.js:
```diff
@@ -274,14 +296,23 @@ define([
-            _.defer(function() {
+            this.model.save().done(function () {
+                spinner.stop();
+                $('#config-modal').modal('hide');
```
You should be able to use the jquery instance that is attached to the view instead of pulling in jquery yourself. Then it is automatically scoped the view instead of the entire DOM. Then you don't need the new "ID" attributes attached to the modals. You can simply do this.$('.modal-content') and it will retrieve your current modal. This will be true for all of these IDs that you added.

pklinef commented:
Please make the error message box look like the insecure defaults alert box at the top of the admin page.  Use a human friendly default message to let the administrator know that there was an error while saving their configuration changes.

brendan-hofmann commented:
Based on conversations with @pklinef and @stustison I'll be refactoring most of this PR. ;) Several topics have come up, including:
 - a more idiomatic marionette approach that drives the view behavior from the model, instead of jquery
 - making the error section look more like the insecure defaults (hopefully through refactoring)
 - fixing the places that use the deprecated deferred callbacks, which are broken now that we're returning deferred directly instead of the jqxhr

brendan-hofmann commented:
I've completed the majority of the changes that were discussed, although it could still use some refinement. I'd also like to handle the progress indicator and logged out case better. I wanted to start getting feedback on the reusable alerts view approach though.

harrison-tarr commented:
Review status: 2 of 16 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[platform/admin/ui/src/main/webapp/js/models/Alerts.js, line 29 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2r0kJoyJhMd4UBNCdx)** ([raw file](https://github.com/codice/ddf/blob/6d2ac0a9c6074fc8ba45594e2019a8e471a19cf4/platform/admin/ui/src/main/webapp/js/models/Alerts.js#L29)):</sup>
This doesn't look too pretty, but also I realize sometimes you gotta do weird stuff. Is this the best/typical way to overwrite the defaults of the parent class?
<!-- Reviewable comment -K2r0kJpUPlGplRt2Cx2:1053888904 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Review status: 2 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[platform/admin/ui/src/main/webapp/templates/alerts.handlebars, line 32 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2r38ShDHbRAhpbgowd)** ([raw file](https://github.com/codice/ddf/blob/6d2ac0a9c6074fc8ba45594e2019a8e471a19cf4/platform/admin/ui/src/main/webapp/templates/alerts.handlebars#L32)):</sup>
what is the difference between `message` and `this.message`?
<!-- Reviewable comment -K2r38ShDHbRAhpbgowe:2032391769 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Review status: 2 of 16 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[platform/admin/ui/src/main/webapp/js/models/Alerts.js, line 29 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2r0kJoyJhMd4UBNCdx)** ([raw file](https://github.com/codice/ddf/blob/6d2ac0a9c6074fc8ba45594e2019a8e471a19cf4/platform/admin/ui/src/main/webapp/js/models/Alerts.js#L29)):</sup>
It is idiomatic of underscore (which backbone is built on) and the js prototype chain, so in this case, I think so. There may be cleaner ways I'm not aware of. ;)
<!-- Reviewable comment -K2w_7IZxhcqQ8ItzU97:-778219885 -->

---

<sup>**[platform/admin/ui/src/main/webapp/templates/alerts.handlebars, line 32 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2r38ShDHbRAhpbgowd)** ([raw file](https://github.com/codice/ddf/blob/6d2ac0a9c6074fc8ba45594e2019a8e471a19cf4/platform/admin/ui/src/main/webapp/templates/alerts.handlebars#L32)):</sup>
Ah, refactoring cruft. `message` is the content of the alert banner, `this.message` is the content of each item in the expandable. I'll clarify it.
<!-- Reviewable comment -K2w_gzHi3nfJmLl7Fkw:-70659154 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Review status: 2 of 16 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[platform/admin/ui/src/main/webapp/templates/alerts.handlebars, line 32 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2r38ShDHbRAhpbgowd)** ([raw file](https://github.com/codice/ddf/blob/6d2ac0a9c6074fc8ba45594e2019a8e471a19cf4/platform/admin/ui/src/main/webapp/templates/alerts.handlebars#L32)):</sup>
oh i think i missed the `each items` bit. so message is a member of the top level JSON object and then this.message is referencing the `message` of `item`?
<!-- Reviewable comment -K2xBeFpzSXPbSLzhvnj:-1795860784 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Review status: 1 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[platform/admin/ui/src/main/webapp/templates/alerts.handlebars, line 32 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2r38ShDHbRAhpbgowd)** ([raw file](https://github.com/codice/ddf/blob/6d2ac0a9c6074fc8ba45594e2019a8e471a19cf4/platform/admin/ui/src/main/webapp/templates/alerts.handlebars#L32)):</sup>
Yes. I've reworded the two values to be clearer. ;)
<!-- Reviewable comment -K2xPzxvQkKPdvPNlDx8:-1283096468 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Manually tested, lgtm

pklinef commented on platform/admin/ui/src/main/webapp/js/models/Service.js:
```diff
@@ -86,6 +86,36 @@ define(['backbone', 'jquery','backboneassociations'],function (Backbone, $) {
+         */
+        handleResult: function (result, deferred) {
+            var jsonResult;
+            var jsonString = result[0];
```
Is it safe to assume that result is an array with at least one element?

pklinef commented on platform/admin/ui/src/main/webapp/js/models/Service.js:
```diff
@@ -86,6 +86,36 @@ define(['backbone', 'jquery','backboneassociations'],function (Backbone, $) {
+                    jsonResult = JSON.parse(jsonString.toString().trim());
+                }
+            } else {
+                jsonResult = {'error': result[1], 'stacktrace': result[2]};
```
Do you need to check the length of result before doing this else?

pklinef commented on platform/admin/ui/src/main/webapp/js/views/Alerts.view.js:
```diff
@@ -0,0 +1,51 @@
+            },
+            serializeData: function () {
+                var json = this.model.toJSON();
+                json.collapseId = parseInt((Math.random() * 10000), 10);
```
The chance of a collision still seems high.  For example, it would be around 11% if there were 50 hidden alert modals.  I would suggest using something like `Math.random() * Math.pow(2, 32)`.

http://www.wolframalpha.com/input/?i=1-e%5E%28%28-50*%2850-1%29%29%2F%282*10000%29%29
http://preshing.com/20110504/hash-collision-probabilities/

brendan-hofmann commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[platform/admin/ui/src/main/webapp/js/models/Service.js, line 95 \[r6\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2JcdKDRi13tsSS-eYA-r6-95)** ([raw file](https://github.com/codice/ddf/blob/6cb299ff58b54458a329111bc7df818d75b18c31/platform/admin/ui/src/main/webapp/js/models/Service.js#L95)):</sup>
It's not even an array, it's actually an Arguments object. ;) Still good to check the length and contents.
<!-- Reviewable comment -K2ycKyjCIWMqlGaWvVb:1007420513 -->

---

<sup>**[platform/admin/ui/src/main/webapp/js/models/Service.js, line 109 \[r6\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2JcdKDRi13tsSS-eYA-r6-109)** ([raw file](https://github.com/codice/ddf/blob/6cb299ff58b54458a329111bc7df818d75b18c31/platform/admin/ui/src/main/webapp/js/models/Service.js#L109)):</sup>
Added some checking.
<!-- Reviewable comment -K2ycS0a59t_9OItwGkU:1836570925 -->

---

<sup>**[platform/admin/ui/src/main/webapp/js/views/Alerts.view.js, line 45 \[r6\]](https://reviewable.io:443/reviews/codice/ddf/314#-K2nlURKSVs5WEwiZqyz-r6-45)** ([raw file](https://github.com/codice/ddf/blob/6cb299ff58b54458a329111bc7df818d75b18c31/platform/admin/ui/src/main/webapp/js/views/Alerts.view.js#L45)):</sup>
Good point, updated.
<!-- Reviewable comment -K2ycVmQuKm0-SEtvvr7:-761667303 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/314)
<!-- Sent from Reviewable.io -->

pklinef commented on platform/admin/ui/src/main/webapp/templates/alerts.handlebars:
```diff
@@ -12,23 +12,30 @@
+
+            </h4>
+        </div>
+        <div class="panel-collapse collapse {{collapse}} collapseAlerts" id="{{collapseId}}">
```
Instead of the id just being a number, I would suggest prefixing the number with something like "alert_".
